File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Generator/GeneratorController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ObjectUtil = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local GetItemMeta = require("Shared/TS/Item/ItemDefinitions").GetItemMeta
local Network = require("Shared/TS/Network").Network
local TimeUtil = require("Shared/TS/Util/TimeUtil").TimeUtil
-- * Generator item spawn offset. Items spawn _above_ generators and fall to the ground.
local GENERATOR_ITEM_SPAWN_OFFSET = Vector3.new(0, 2.25, 0)
-- * Generator label offset. Labels are _above_ the item spawn location.
local GENERATOR_LABEL_OFFSET = Vector3.new(0, 3, 0)
local GeneratorController
do
	GeneratorController = setmetatable({}, {
		__tostring = function()
			return "GeneratorController"
		end,
	})
	GeneratorController.__index = GeneratorController
	function GeneratorController.new(...)
		local self = setmetatable({}, GeneratorController)
		return self:constructor(...) or self
	end
	function GeneratorController:constructor()
		self.generatorMap = {}
		self.spawnResetGenerators = {}
		self.stackedGenerators = {}
		self.generatorTextLabelMap = {}
		-- Set up generator item collision rules.
		Physics:IgnoreLayerCollision(3, 12)
		Physics:IgnoreLayerCollision(12, 12)
		-- NOTE: This is temp. Prefab will be dynamically loaded per generator.
		self.generatorItemPrefab = AssetBridge:LoadAsset("Shared/Resources/Prefabs/GeneratorItemPlaceholder.prefab")
		-- NOTE: Placeholder label.
		self.generatorLabelPrefab = AssetBridge:LoadAsset("Client/Resources/Prefabs/GeneratorLabel.prefab")
	end
	function GeneratorController:OnStart()
		-- Listen for generator snapshot. Should only be received on late joins.
		Network.ServerToClient.GeneratorSnapshot.Client:OnServerEvent(function(generatorStateDtos)
			local _generatorStateDtos = generatorStateDtos
			local _arg0 = function(dto)
				-- Skip generator if it already exists on client.
				local _generatorMap = self.generatorMap
				local _id = dto.id
				if _generatorMap[_id] ~= nil then
					return nil
				end
				-- Otherwise, create.
				local timeUntilNextSpawn = dto.nextSpawnTime - TimeUtil:GetServerTime()
				if timeUntilNextSpawn > 0 then
					dto.spawnRate = timeUntilNextSpawn
					local _spawnResetGenerators = self.spawnResetGenerators
					local _id_1 = dto.id
					_spawnResetGenerators[_id_1] = true
				end
				local _generatorMap_1 = self.generatorMap
				local _id_1 = dto.id
				local _dto = dto
				_generatorMap_1[_id_1] = _dto
				if dto.label then
					self:CreateGeneratorLabel(dto)
				end
			end
			for _k, _v in _generatorStateDtos do
				_arg0(_v, _k - 1, _generatorStateDtos)
			end
		end)
		-- Listen for generator creation.
		Network.ServerToClient.GeneratorCreated.Client:OnServerEvent(function(dto)
			-- Adjust initial spawn time to sync with server.
			local timeUntilNextSpawn = dto.nextSpawnTime - TimeUtil:GetServerTime()
			if timeUntilNextSpawn > 0 then
				dto.spawnRate = timeUntilNextSpawn
				local _spawnResetGenerators = self.spawnResetGenerators
				local _id = dto.id
				_spawnResetGenerators[_id] = true
			end
			local _generatorMap = self.generatorMap
			local _id = dto.id
			local _dto = dto
			_generatorMap[_id] = _dto
			-- Set up generator label if applicable.
			if dto.label then
				self:CreateGeneratorLabel(dto)
			end
		end)
		-- Listen for generator looted.
		Network.ServerToClient.GeneratorLooted.Client:OnServerEvent(function(generatorId)
			local _generatorMap = self.generatorMap
			local _generatorId = generatorId
			local dto = _generatorMap[_generatorId]
			if not dto then
				return nil
			end
			-- Update generator label if applicable.
			if dto.label then
				self:UpdateGeneratorTextLabel(dto.id)
			end
			-- Delete generator root GameObject.
			local _stackedGenerators = self.stackedGenerators
			local _id = dto.id
			local rootGO = _stackedGenerators[_id]
			if rootGO then
				local _stackedGenerators_1 = self.stackedGenerators
				local _id_1 = dto.id
				_stackedGenerators_1[_id_1] = nil
				GameObjectBridge:Destroy(rootGO)
			end
		end)
	end
	function GeneratorController:CreateGeneratorLabel(dto)
		local labelPosition = dto.pos + GENERATOR_LABEL_OFFSET
		local generatorLabel = GameObjectBridge:InstantiateAt(self.generatorLabelPrefab, labelPosition, Quaternion.identity)
		-- Set initial label text.
		local generatorTextTransform = generatorLabel.transform:FindChild("GeneratorText")
		local generatorTextComponent = generatorTextTransform:GetComponent("TextMeshProUGUI")
		local itemMeta = GetItemMeta(dto.item)
		generatorTextComponent.text = itemMeta.displayName .. " Generator"
		local _generatorTextLabelMap = self.generatorTextLabelMap
		local _id = dto.id
		_generatorTextLabelMap[_id] = generatorTextComponent
	end
	function GeneratorController:UpdateGeneratorTextLabel(generatorId)
		local _generatorMap = self.generatorMap
		local _generatorId = generatorId
		local dto = _generatorMap[_generatorId]
		if not dto then
			return nil
		end
		local _generatorTextLabelMap = self.generatorTextLabelMap
		local _generatorId_1 = generatorId
		local generatorTextComponent = _generatorTextLabelMap[_generatorId_1]
		if not generatorTextComponent then
			return nil
		end
		local itemMeta = GetItemMeta(dto.item)
		generatorTextComponent.text = itemMeta.displayName .. " Generator"
	end
	function GeneratorController:GetAllGenerators()
		return ObjectUtil.values(self.generatorMap)
	end
end
-- (Flamework) GeneratorController metadata
Reflect.defineMetadata(GeneratorController, "identifier", "Bundles/Client/Controllers/Global/Generator/GeneratorController@GeneratorController")
Reflect.defineMetadata(GeneratorController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(GeneratorController, "$:flamework@Controller", Controller, { {} })
return {
	GeneratorController = GeneratorController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Client/Controllers/Generator/GeneratorController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import ObjectUtil from "@easy-games/unity-object-utils";
import { CoreNetwork } from "Shared/CoreNetwork";
import { GameObjectUtil } from "Shared/GameObject/GameObjectUtil";
import { GeneratorDto } from "Shared/Generator/GeneratorMeta";
import { ItemUtil } from "Shared/Item/ItemUtil";
import { Layer } from "Shared/Util/Layer";
import { TimeUtil } from "Shared/Util/TimeUtil";

/** Generator item spawn offset. Items spawn _above_ generators and fall to the ground. */
const GENERATOR_ITEM_SPAWN_OFFSET = new Vector3(0, 2.25, 0);
/** Generator label offset. Labels are _above_ the item spawn location.*/
const GENERATOR_LABEL_OFFSET = new Vector3(0, 3, 0);

@Controller({})
export class GeneratorController implements OnStart {
	/** Generator item prefab. */
	private generatorItemPrefab: Object;
	/** Generator label prefab. */
	private generatorLabelPrefab: Object;
	/** Mapping of generator id to `GeneratorStateDto`. */
	private generatorMap = new Map<string, GeneratorDto>();
	/** Set of generators that require a spawn time reset. */
	private spawnResetGenerators = new Set<string>();
	/**
	 * Map of generators to stack root GameObject. Indicates whether or not client should delete
	 * generator dropped items.
	 */
	private stackedGenerators = new Map<string, GameObject>();
	/** Map of generators to text label components. */
	private generatorTextLabelMap = new Map<string, TextMeshProUGUI>();

	constructor() {
		/* Set up generator item collision rules. */
		Physics.IgnoreLayerCollision(Layer.CHARACTER, Layer.GENERATOR_ITEM);
		Physics.IgnoreLayerCollision(Layer.GENERATOR_ITEM, Layer.GENERATOR_ITEM);
		/* NOTE: This is temp. Prefab will be dynamically loaded per generator. */
		this.generatorItemPrefab = AssetBridge.LoadAsset("Shared/Resources/Prefabs/GeneratorItemPlaceholder.prefab");
		/* NOTE: Placeholder label. */
		this.generatorLabelPrefab = AssetBridge.LoadAsset("Client/Resources/Prefabs/GeneratorLabel.prefab");
	}

	OnStart(): void {
		/* Listen for generator snapshot. Should only be received on late joins. */
		CoreNetwork.ServerToClient.GeneratorSnapshot.Client.OnServerEvent((generatorStateDtos) => {
			generatorStateDtos.forEach((dto) => {
				/* Skip generator if it already exists on client. */
				if (this.generatorMap.has(dto.id)) return;
				/* Otherwise, create. */
				const timeUntilNextSpawn = dto.nextSpawnTime - TimeUtil.GetServerTime();
				if (timeUntilNextSpawn > 0) {
					dto.spawnRate = timeUntilNextSpawn;
					this.spawnResetGenerators.add(dto.id);
				}
				this.generatorMap.set(dto.id, dto);

				if (dto.label) {
					this.CreateGeneratorLabel(dto);
				}
			});
		});
		/* Listen for generator creation. */
		CoreNetwork.ServerToClient.GeneratorCreated.Client.OnServerEvent((dto) => {
			/* Adjust initial spawn time to sync with server. */
			const timeUntilNextSpawn = dto.nextSpawnTime - TimeUtil.GetServerTime();
			if (timeUntilNextSpawn > 0) {
				dto.spawnRate = timeUntilNextSpawn;
				this.spawnResetGenerators.add(dto.id);
			}
			this.generatorMap.set(dto.id, dto);
			/* Set up generator label if applicable. */
			if (dto.label) this.CreateGeneratorLabel(dto);
		});
		/* Listen for generator looted. */
		CoreNetwork.ServerToClient.GeneratorLooted.Client.OnServerEvent((generatorId) => {
			const dto = this.generatorMap.get(generatorId);
			if (!dto) return;
			/* Update generator label if applicable. */
			if (dto.label) this.UpdateGeneratorTextLabel(dto.id);
			/* Delete generator root GameObject. */
			const rootGO = this.stackedGenerators.get(dto.id);
			if (rootGO) {
				this.stackedGenerators.delete(dto.id);
				GameObjectUtil.Destroy(rootGO);
			}
		});
	}

	/** Creates a generator label in world space. */
	private CreateGeneratorLabel(dto: GeneratorDto): void {
		const labelPosition = dto.pos.add(GENERATOR_LABEL_OFFSET);
		const generatorLabel = GameObjectUtil.InstantiateAt(
			this.generatorLabelPrefab,
			labelPosition,
			Quaternion.identity,
		);
		/* Set initial label text. */
		const generatorTextTransform = generatorLabel.transform.FindChild("GeneratorText")!;
		const generatorTextComponent = generatorTextTransform.GetComponent<TextMeshProUGUI>();
		const itemMeta = ItemUtil.GetItemMeta(dto.item);
		generatorTextComponent.text = `${itemMeta.displayName} Generator`;
		this.generatorTextLabelMap.set(dto.id, generatorTextComponent);
	}

	/** Update a generator text label. */
	private UpdateGeneratorTextLabel(generatorId: string): void {
		const dto = this.generatorMap.get(generatorId);
		if (!dto) return;
		const generatorTextComponent = this.generatorTextLabelMap.get(generatorId);
		if (!generatorTextComponent) return;

		const itemMeta = ItemUtil.GetItemMeta(dto.item);
		generatorTextComponent.text = `${itemMeta.displayName} Generator`;
	}

	/**
	 * @returns All `GeneratorStateDto`s on client.
	 */
	public GetAllGenerators(): GeneratorDto[] {
		return ObjectUtil.values(this.generatorMap);
	}
}
